### PR TITLE
add custom command to report and an option to log only custom commands

### PR DESCRIPTION
**Pull request by FelixZilber**
Added an option for custom steps for report and ability to log only custom commands to reporter for cleaner look.

Usage: 
`  process.send({
            event: 'runner:customCommand',
            title: `step title`,
            body: `step body as an attachment`,
            type: 'test/plain'
        });`

in order to get only custom logs need to add parameter in wdio.config.js:
`reporters: ['dot', 'allure'],
    reporterOptions: {
        allure: {
            outputDir: 'allure-results',
            **customStepsOnly: true,**
        },
    },`

### CODE REVIEW

File: lib/reporter.js [hunk 1]
```diff
@@ -108,6 +109,9 @@ class AllureReporter extends events.EventEmitter {
     }
 
     runnerCommand (command) {
+        if (!this.options.customStepsOnly) {
```
Comment (christian-bromann) [inline lib/reporter.js]: It would be great if this option could get documented

**Comment by FelixZilber** [comment]
Attaching screenshots with custom commands only and with all commands:
testwithallcommands
testwithcustomcommandsonly

File: lib/reporter.js (hunk 1 above)
Comment (FelixZilber) [reply lib/reporter.js]: Added example to README

**Comment by BorisOsipov** [comment]
@FelixZilber
 It is a nice idea to implement creating step from the test. Good thing for reporter users 👍

Let me give you some advice to you in implementation.

It will be better and flexible if you
- create `reporter.createStep(data)` method and allow you and other users to call it from his code to create any step in test (in your case it steps for custom commands) rather than use process.send directly.
See for example implementation
https://github.com/webdriverio/wdio-allure-reporter/blob/6aedcd4459f5ac0d092cc197681b3d7b53520d01/lib/reporter.js#L210
- think about the API. Especially allure steps fields. They are at least 'title', 'name' - but you fill only title.
- rename `customStepsOnly` => `disableWebdriverStepsReporting` - it more meaningful

If you need help\have question feel free to ask me.

**Comment by just-boris** [comment]
@FelixZilber instead of adding a conditional behavior here, you may also consider using [allure-mocha](https://github.com/allure-framework/allure-mocha/) (there are also similar packages for other frameworks). It doesn't emit any information by default, leaving this up to you: features/stories definition, steps, attachments.

For example, here we have an example of Mocha + Allure + WebdriverIO setup: https://github.com/allure-examples/mocha-allure-example.

**Comment by FelixZilber** [comment]
@just-boris using allure-mocha requires more work like failure handling, while with wdio-allure-reporter it just about output customization.

**Comment by FelixZilber** [comment]
For some reason ALL tests failing locally(npm run test). 
I got:      Error: ENOENT: no such file or directory, scandir '/Users/felixzilber/dev/wdio-allure-reporter/.allure-results'

What am I missing?

**Comment by BorisOsipov** [comment]
Have you ran `./node_modules/.bin/static ./test/fixtures` to serve test pages? 
It desccibed in
https://github.com/webdriverio/wdio-allure-reporter#development

**Comment by just-boris** [comment]
> allure-mocha requires more work like failure handling
If you are talking about capability to take the bailing out screenshot in case of error, it is solved there via [simple one-liner](https://github.com/allure-examples/mocha-allure-example/blob/master/test/webdriver-io.spec.js#L105).

If you have anything else, I would be happy to know as an issue to the respective repo.

**Comment by christian-bromann** [comment]
@just-boris @BorisOsipov @FelixZilber what is the state of this PR? Can we merge it or are there any outstanding issues? A little bit hard to tell according to the discussions above.

**Comment by FelixZilber** [comment]
@christian-bromann  There is a failed test, that I need to fix once I'll manage to run it locally

File: lib/reporter.js [hunk 2]
```diff
@@ -40,6 +40,7 @@ class AllureReporter extends events.EventEmitter {
 
         // Runner events (webdriver)
         this.on('runner:command', ::this.runnerCommand)
+        this.on('runner:customCommand', ::this.customCommand)
```
Comment (BorisOsipov) [inline lib/reporter.js]: @FelixZilber 
Here and below. let's rename `customCommand` to `step`.
I will be more meaningful

Comment (BorisOsipov) [reply lib/reporter.js]: what do you think?

Comment (FelixZilber) [reply lib/reporter.js]: Agree. Pushed the change

**Comment by FelixZilber** [comment]
@BorisOsipov @just-boris @christian-bromann  I guess we are good to go

**Comment by christian-bromann** [comment]
@FelixZilber can we have some tests for this?

**Comment by FelixZilber** [comment]
@christian-bromann  
Test added. It passed locally:
```
node --version
v9.3.0

npm --version
5.6.0
```
 Step
    ✓ should add step to the test (1653ms)
    ✓ should have attachment (1691ms)


But failed in travis. 

Ideas?

**Comment by BorisOsipov** [comment]
@FelixZilber can you rebase your branch to `master`? Let see may it help

File: test/specs/step.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { expect } from 'chai'
+import { clean, run, getResultFiles, getResultFileValue } from '../helper'
+
+describe('Step', () => {
+    beforeEach(clean)
+
+    it('should add step to the test', () => {
+        return run(['step']).then((results) => {
```
Comment (BorisOsipov) [inline test/specs/step.js]: Tests aren't checking that in the corresponding `test case` has created `step`

Comment (FelixZilber) [reply test/specs/step.js]: added more tests

**Comment by FelixZilber** [comment]
@BorisOsipov some thing went wrong after rebase with master

**Comment by BorisOsipov** [comment]
Restart travis

**Comment by christian-bromann** [comment]
@FelixZilber can you create a new branch and cherry-pick only these commits that are actual relevant? I don't think we can merge this PR anymore due to the rebase issues.

**Comment by FelixZilber** [comment]
will create new PR

**Comment by sumeetp4** [comment]
Hey @FelixZilber , how is this createStep function different createStep in mocha-allure API?

**Comment by sumeetp4** [comment]
hey @BorisOsipov , 
 Given-When-Then in my scenarios are considered as individual test cases. How do I configure them to be as steps of a single test case (single scenario)

**Comment by BorisOsipov** [comment]
@sumeetp4
see https://github.com/webdriverio/wdio-allure-reporter#configuration
`useCucumberStepReporter`

**Comment by sumeetp4** [comment]
Thanks @BorisOsipov , it works like a charm :+1:

**Comment by sumeetp4** [comment]
@BorisOsipov ,

Also any idea why this test case is coming undefined and the sub steps are coming as unknown?

allure

**Comment by sumeetp4** [comment]
Okay sure :+1:

**Pull request closed without merging**